docker/docker_client.go

```diff
@@ -181,13 +184,27 @@ func newDockerClientFromRef(sys *types.SystemContext, ref dockerReference, write
 	if err != nil {
 		return nil, err
 	}
-	remoteName := reference.Path(ref.ref)
 
-	return newDockerClientWithDetails(sys, registry, username, password, actions, sigBase, remoteName)
+	client, err := newDockerClient(sys, registry, ref.ref.Name())
+	if err != nil {
+		return nil, err
+	}
+	client.username = username
+	client.password = password
+	client.signatureBase = sigBase
+	client.scope.actions = actions
+	client.scope.remoteName = reference.Path(ref.ref)
+	return client, nil
 }
 
-// newDockerClientWithDetails returns a new dockerClient instance for the given parameters
-func newDockerClientWithDetails(sys *types.SystemContext, registry, username, password, actions string, sigBase signatureStorageBase, remoteName string) (*dockerClient, error) {
+// newDockerClient returns a new dockerClient instance for the given registry
+// and reference.  The reference is used to query the registry configuration
+// and can either be registry (.e.g, "registry.com:5000"), a fully-qualified
+// image reference (e.g., "docker.io/library/alpine:latest") or pointing to a
```

Comment:
Wait… this actually works… because `FindRegistry` just uses `strings.HasPrefix` and does not require the matched `Prefix` to be followed by any kind of semantic separator (like `/` or maybe `:`.

Thus, `Prefix` `example.com/1` matches `example.com/1/foo` as well as `example.com/100/foo`. Is that what we want?

Just below the comment goes on discussing a registries namespace — but a string prefix and a namespace are not the same thing.

(The `strings.HasPrefix` match has existed before this PR, so this is not blocking, I can split it into a separate PR for discussion.)

Reply:
> Thus, Prefix example.com/1 matches example.com/1/foo as well as example.com/100/foo. Is that what we want?
> Just below the comment goes on discussing a registries namespace — but a string prefix and a namespace are not the same thing.

That's definitely an issue. Given our timezones are quite distant, I have enough time and can add a commit ontop of this PR.

Reply:
> That's definitely an issue. Given our timezones are quite distant, I have enough time and can add a commit ontop of this PR.

@mtrmac I added another commit. A prefix match is now limited to the following cases:

- full match (`ref, prefix = "docker.io/alpine:latest"`)
- namespace match (`ref = "docker.io/alpine/namespace"` and `prefix = "docker.io/alpine"`)
  - the remainder of `ref` starts with a `/`
- image match (`ref = "docker.io/alpine:latest"` and `prefix = "docker.io/alpine"`)
  - the remainder of `ref` starts with a `:`

Reply:
^ a namespace match certainly also applies to matching a registry (e.g., `prefix = "registry.com:5000")

Reply:
This is an internal API; so please document only the two actually necessary inputs: either only a hostname[:port], or a hostname:[port]/[ns/…]/repo ; with hostname[:port] a single unit that must match exactly (although the implementation may do more than that for now).

(This does not directly affect the implementation, but avoids even a pretend commitment to support unnecessary formats, making future refactoring easier.)

Reply:
Done.